File: app/src/main/java/com/hippo/ehviewer/client/parser/ArchiveParser.java
/*
 * Copyright 2016 Hippo Seven
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.hippo.ehviewer.client.parser;

import androidx.annotation.NonNull;

import java.util.ArrayList;
import java.util.List;
import java.util.Objects;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public class ArchiveParser {

    private static final Pattern PATTERN_FORM = Pattern.compile("<form id=\"hathdl_form\" action=\"[^\"]*?or=([^=\"]*?)\" method=\"post\">");
    private static final Pattern PATTERN_ARCHIVE = Pattern.compile("<p><a href=\"[^\"]*\" onclick=\"return do_hathdl\\('([0-9]+|org)'\\)\">([^<]+)</a></p>\\s*<p>([\\w. ]+)</p>\\s*<p>([\\w. ]+)</p>");

    public static Result parse(String body) {
        Matcher m = PATTERN_FORM.matcher(body);
        if (!m.find()) {
            return new Result(null, null);
        }
        String paramOr = m.group(1);
        var archiveList = new ArrayList<Archive>();
        m = PATTERN_ARCHIVE.matcher(body);
        while (m.find()) {
            String res = ParserUtils.trim(m.group(1));
            String name = ParserUtils.trim(m.group(2));
            String size = ParserUtils.trim(m.group(3));
            String cost = ParserUtils.trim(m.group(4));
            var item = new Archive(res, name, size, cost);
            archiveList.add(item);
        }
        return new Result(paramOr, archiveList);
    }

    public static final class Archive {
        private final String res;
        private final String name;
        private final String size;
        private final String cost;

        public Archive(String res, String name, String size, String cost) {
            this.res = res;
            this.name = name;
            this.size = size;
            this.cost = cost;
        }

        public String format() {
            return String.format("%s [%s, %s]", name, size, cost);
        }

        public String res() {
            return res;
        }

        public String name() {
            return name;
        }

        public String size() {
            return size;
        }

        public String cost() {
            return cost;
        }

        @Override
        public boolean equals(Object obj) {
            if (obj == this) return true;
            if (obj == null || obj.getClass() != this.getClass()) return false;
            var that = (Archive) obj;
            return Objects.equals(this.res, that.res) &&
                    Objects.equals(this.name, that.name) &&
                    Objects.equals(this.size, that.size) &&
                    Objects.equals(this.cost, that.cost);
        }

        @Override
        public int hashCode() {
            return Objects.hash(res, name, size, cost);
        }

        @NonNull
        @Override
        public String toString() {
            return "Archive[" +
                    "res=" + res + ", " +
                    "name=" + name + ", " +
                    "size=" + size + ", " +
                    "cost=" + cost + ']';
        }

    }

    public static final class Result {
        private final String paramOr;
        private final List<Archive> archiveList;

        public Result(String paramOr, List<Archive> archiveList) {
            this.paramOr = paramOr;
            this.archiveList = archiveList;
        }

        public String paramOr() {
            return paramOr;
        }

        public List<Archive> archiveList() {
            return archiveList;
        }

        @Override
        public boolean equals(Object obj) {
            if (obj == this) return true;
            if (obj == null || obj.getClass() != this.getClass()) return false;
            var that = (Result) obj;
            return Objects.equals(this.paramOr, that.paramOr) &&
                    Objects.equals(this.archiveList, that.archiveList);
        }

        @Override
        public int hashCode() {
            return Objects.hash(paramOr, archiveList);
        }

        @NonNull
        @Override
        public String toString() {
            return "Result[" +
                    "paramOr=" + paramOr + ", " +
                    "archiveList=" + archiveList + ']';
        }

    }
}

File: app/src/main/java/com/hippo/ehviewer/client/parser/ArchiveParser.kt
/*
 * Copyright 2016 Hippo Seven
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.hippo.ehviewer.client.parser

import com.hippo.ehviewer.client.exception.NoHAtHClientException
import org.jsoup.Jsoup

object ArchiveParser {
    private val PATTERN_ARCHIVE_URL =
        Regex("<strong>(.*)</strong>.*<a href=\"([^\"]*)\">Click Here To Start Downloading</a>")
    private val PATTERN_CURRENT_FUNDS =
        Regex("<p>([\\d,]+) GP \\[[^]]*] &nbsp; ([\\d,]+) Credits \\[[^]]*]</p>")
    private val PATTERN_FUNDS =
        Regex("Available: ([\\d,]+) Credits.*Available: ([\\d,]+) kGP", RegexOption.DOT_MATCHES_ALL)
    private val PATTERN_HATH_FORM =
        Regex("<form id=\"hathdl_form\" action=\"[^\"]*?or=([^=\"]*?)\" method=\"post\">")
    private val PATTERN_HATH_ARCHIVE =
        Regex("<p><a href=\"[^\"]*\" onclick=\"return do_hathdl\\('([0-9]+|org)'\\)\">([^<]+)</a></p>\\s*<p>([\\w. ]+)</p>\\s*<p>([\\w. ]+)</p>")
    private val PATTERN_NEED_HATH_CLIENT =
        Regex("You must have a H@H client assigned to your account to use this feature\\.")

    fun parse(body: String): Result? {
        val m = PATTERN_HATH_FORM.find(body) ?: return null
        val paramOr = m.groupValues[1]
        val archiveList = ArrayList<Archive>()
        Jsoup.parse(body).select("#db>div>div").forEach { element ->
            if (element.childrenSize() > 0 && !element.attr("style").contains("color:#CCCCCC")) {
                runCatching {
                    val res = element.selectFirst("form>input")!!.attr("value")
                    val name = element.selectFirst("form>div>input")!!.attr("value")
                    val size = element.selectFirst("p>strong")!!.text()
                    val cost = element.selectFirst("div>strong")!!.text().replace(",", "")
                    Archive(res, name, size, cost, false)
                }.onSuccess {
                    archiveList.add(it)
                }.onFailure {
                    it.printStackTrace()
                }
            }
        }
        PATTERN_HATH_ARCHIVE.findAll(body).forEach { matchResult ->
            val (res, name, size, cost) = matchResult.groupValues.slice(1..4)
                .map { ParserUtils.trim(it) }
            val item = Archive(res, name, size, cost, true)
            archiveList.add(item)
        }
        val result = Result(paramOr, archiveList, null)
        PATTERN_CURRENT_FUNDS.find(body)?.groupValues?.run {
            val fundsGP = ParserUtils.parseInt(get(1), 0)
            val fundsC = ParserUtils.parseInt(get(2), 0)
            val funds = Funds(fundsGP, fundsC)
            result.funds = funds
        }
        return result
    }

    @Throws(NoHAtHClientException::class)
    fun parseArchiveUrl(body: String): String? {
        if (PATTERN_NEED_HATH_CLIENT.containsMatchIn(body)) throw NoHAtHClientException("No H@H client")
        return Jsoup.parse(body).selectFirst("#continue>a[href]")
            ?.let { it.attr("href") + "?start=1" }
        // TODO: Check more errors
    }

    fun parseFunds(body: String): Funds? {
        return PATTERN_FUNDS.find(body)?.groupValues?.run {
            val fundsC = ParserUtils.parseInt(get(1), 0)
            val fundsGP = ParserUtils.parseInt(get(2), 0) * 1000
            return Funds(fundsGP, fundsC)
        }
    }

    class Archive(
        val res: String,
        val name: String,
        val size: String,
        val cost: String,
        val isHAtH: Boolean
    )

    class Funds(var fundsGP: Int, var fundsC: Int)

    class Result(val paramOr: String?, val archiveList: List<Archive>, var funds: Funds?)
}
